apps/functions-subscription/src/use-cases/insert-subscription.ts

```diff
@@ -0,0 +1,69 @@
+import { pipe } from 'fp-ts/lib/function';
+import * as O from 'fp-ts/Option';
+import * as TE from 'fp-ts/TaskEither';
+import * as RTE from 'fp-ts/ReaderTaskEither';
+import { Capabilities } from '../domain/capabilities';
+import {
+  SubscriptionId,
+  TrialId,
+  UserId,
+  makeSubscriptionId,
+} from '../domain/subscription';
+import { SubscriptionAlreadyExists, SubscriptionStoreError } from './errors';
+import { nowDate } from '../domain/clock';
+
+// Maps all the requirements for this use-case
+type Env = Pick<
+  Capabilities,
+  | 'subscriptionRequestWriter'
+  | 'subscriptionReader'
+  | 'subscriptionWriter'
+  | 'hashFn'
+  | 'clock'
+>;
+
+const handleSubscriptionAlreadyExists =
+  (id: SubscriptionId) => (env: Pick<Capabilities, 'subscriptionReader'>) =>
+    pipe(
+      env.subscriptionReader.get(id),
+      TE.flatMap(
+        O.fold(
+          () => TE.right(id),
+          () =>
+            TE.left(
+              new SubscriptionAlreadyExists('Subscription already exists'),
+            ),
+        ),
+      ),
+    );
```

Comment:
```suggestion
    pipe(
      env.subscriptionReader.get(id),
      TE.chainW(
       TE.fromOption(() => new SubscriptionAlreadyExists('Subscription already exists'))
      ),
      TE.map(() => id),
    );
```

Reply:
I like this suggestion, it simplifies the code! But we need the opposite:
- return the `id` if the `Option` is `None`
- return the `Error` if the `Option` is `Some`

Reply:
Maybe we can refactor with:
```
env.subscriptionReader.get(id),
      TE.chainW(
        flow(
          TE.fromOption(() => id),
          TE.swap,
          TE.mapLeft(new SubscriptionAlreadyExists('Subscription already exists'))
        )
      )
```